src/Analyser/MutatingScope.php

```diff
@@ -4406,6 +4406,26 @@ public function assignInitializedProperty(Type $fetchedOnType, string $propertyN
 		return $this->assignExpression(new PropertyInitializationExpr($propertyName), new MixedType(), new MixedType());
 	}
 
+	/**
+	 * @param array<string, ExpressionTypeHolder> $expressionTypes
+	 *
+	 * @return array<string, ExpressionTypeHolder>
+	 */
+	private function preserveResolvedTypes(array $expressionTypes): array
+	{
+		$preservedTypes = $this->resolvedTypes;
+		foreach($preservedTypes as $exprStringToInvalidate => $resolvedType) {
+			foreach ($expressionTypes as $exprString => $exprTypeHolder) {
```

Comment:
we cannot use similar logic as in `invalidateExpression` because we don't know the `Expr` which `$this->resolvedTypes` was resolved with

Reply:
I have a commit that puts ExpressionTypeHolder into resolvedTypes, do you want it?

Reply:
I pushed it to 2.1.x. If it doesn't help you in the end, or if this PR does not lead to performance improvements, I'll revert it.